frostsnap_core/src/coordinator.rs

```diff
@@ -1170,93 +1178,192 @@ impl FrostCoordinator {
+            .get_access_structure(held_share.access_structure_ref)
+            .is_none());
+        self.mutate(Mutation::NewRestoration {
+            restoration_id,
+            key_name: held_share.key_name,
+            threshold: held_share.threshold,
+            key_purpose: held_share.purpose,
+        });
 
-        let coord_key = self
-            .keys
-            .get(&access_structure_ref.key_id)
-            .expect("mutation should have created it");
+        self.mutate(Mutation::RestorationProgress {
+            restoration_id,
+            device_id: recover_share.held_by,
+            access_structure_ref: Some(held_share.access_structure_ref),
+            share_image: held_share.share_image,
+        });
+    }
 
-        // This may be `None` if the access structure has already been recovered
-        if let Some(pending_as) = coord_key
-            .recovering_access_structures
-            .get(&access_structure_ref.access_structure_id)
-        {
-            return pending_as.share_images.len() >= pending_as.threshold as usize;
-        }
+    pub fn cancel_restoration(&mut self, restoration_id: RestorationId) {
+        self.mutate(Mutation::CancelRestoration { restoration_id })
+    }
+
+    pub fn start_restore_physical_backup(
+        &self,
+        restoration_id: RestorationId,
+        device_id: DeviceId,
+    ) -> impl IntoIterator<Item = CoordinatorSend> {
+        vec![CoordinatorSend::ToDevice {
+            message: CoordinatorToDeviceMessage::LoadPhysicalBackup { restoration_id },
+            destinations: [device_id].into(),
+        }]
+    }
 
-        false
+    pub fn finish_restore_physical_backup(&mut self, phase: PhysicalBackupPhase) {
+        self.mutate(Mutation::RestorationProgress {
+            restoration_id: phase.backup.restoration_id,
+            device_id: phase.from,
+            access_structure_ref: None,
+            share_image: phase.backup.share_image,
+        });
     }
 
-    pub fn recover_share_and_maybe_recover_access_structure(
+    pub fn add_recovery_share_to_restoration(
         &mut self,
+        restoration_id: RestorationId,
         recover_share: RecoverShare,
-        encryption_key: SymmetricKey,
-        rng: &mut impl rand_core::RngCore,
-    ) -> Result<(), ActionError> {
-        let access_structure_ref = recover_share.held_share.access_structure_ref;
-        let ready_for_as_recovery = self.recover_share(recover_share);
+    ) -> Result<(), RestoreRecoverShareError> {
+        self.check_recover_share_compatible_with_restoration(restoration_id, &recover_share)?;
+        self.mutate(Mutation::RestorationProgress {
+            restoration_id,
+            device_id: recover_share.held_by,
+            access_structure_ref: Some(recover_share.held_share.access_structure_ref),
+            share_image: recover_share.held_share.share_image,
+        });
+
+        Ok(())
+    }
+
+    pub fn check_recover_share_compatible_with_restoration(
```

Comment:
review todo: check if we stop at T, or, what happens at T+1

Reply:
Continuing beyond `t` is always useful. Hopefully we can automatically tell that one of the shares is invalid when you have `t`. Adding more (correct) shares can tell you which one. The polynomial id thing is of course ideal for distinguishing related and unrelated shares but that doesn't seem to be crucial to the outcome of the restoration.